discord/automod.py

```diff
@@ -0,0 +1,398 @@
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Optional, List, Union, Type
+
+
+from .enums import AutoModRuleTriggerType, AutoModRuleActionType, AutoModRuleEventType, try_enum
+from .flags import AutoModPresets
+from . import utils
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+    from .abc import GuildChannel
+    from .threads import Thread
+    from .guild import Guild
+    from .member import Member
+    from .state import ConnectionState
+    from .types.automod import (
+        AutoModerationRule as AutoModerationRulePayload,
+        AutoModerationTriggerMetadata as AutoModerationTriggerMetadataPayload,
+        AutoModerationAction as AutoModerationActionPayload,
+        AutoModerationActionExecution as AutoModerationActionExecutionPayload,
+    )
+    from .user import User
+    from .role import Role
+
+__all__ = (
+    'AutoModRuleAction',
+    'AutoModTrigger',
+    'AutoModRule',
+    'AutoModRuleExecution',
+)
+
+
+class AutoModRuleAction:
+    """Represents a rule action.
+
+    Attributes
+    -----------
+    type: :class:`AutoModRuleActionType`
+        The type of action to take.
+    channel_id: Optional[int]
+        The channel to send the alert message to, if any.
+    duration_seconds: Optional[int]
+        The duration of the timeout to apply, if any.
+
+    Raises
+    -------
+    :exc:`ValueError`
+        Provided arguments do not match the needed types.
+    """
+
+    __slots__ = ('type', 'channel_id', 'duration_seconds')
+
+    def __init__(
+        self, *, type: AutoModRuleActionType, channel_id: Optional[int] = None, duration_seconds: Optional[int] = None
+    ) -> None:
+        self.type: AutoModRuleActionType = type
+        self.channel_id: Optional[int] = channel_id
+        self.duration_seconds: Optional[int] = duration_seconds
+
+        if self.type is AutoModRuleActionType.timeout and self.duration_seconds is None:
+            raise ValueError('duration_seconds must be set if type is timeout.')
+        elif self.type is AutoModRuleActionType.send_alert_message and self.channel_id is None:
+            raise ValueError('channel_id must be set if type is send_alert_message.')
+
+    @classmethod
+    def from_data(cls: Type[Self], data: AutoModerationActionPayload) -> Self:
```

Comment:
You don't need Type[Self] as an annotation here
```suggestion
    def from_data(cls, data: AutoModerationActionPayload) -> Self:
```